clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/client/P2PFileTransferClientHandler.java

```diff
@@ -129,13 +170,45 @@ protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws Ex
               "File size mismatch for " + fileName + ". Expected: " + fileContentLength + ", Actual: "
                   + outputFileChannel.size());
         }
-        resetState();
       }
     } else {
       throw new VeniceException("Unexpected message received: " + msg.getClass().getName());
     }
   }
 
+  /**
+   * Handles the metadata transfer response.
+   * @param ctx the channel context
+   * @param msg the http object
+   * @throws JsonProcessingException
+   */
+  private BlobTransferPartitionMetadata handlerMetadataTransferResponse(ChannelHandlerContext ctx, HttpObject msg)
+      throws JsonProcessingException {
+    if (msg instanceof HttpContent) {
+      HttpContent httpContent = (HttpContent) msg;
+      ByteBuf content = httpContent.content();
+      if (content.readableBytes() == 0) {
+        LOGGER.info("Empty content received for metadata transfer");
+        return null;
+      }
+      if (content.isReadable()) {
+        accumulatedMetadataContent.writeBytes(content);
+      }
+      if (msg instanceof DefaultLastHttpContent) {
+        if (accumulatedMetadataContent.isReadable()) {
+          byte[] contentBytes = new byte[accumulatedMetadataContent.readableBytes()];
+          accumulatedMetadataContent.readBytes(contentBytes);
+          ObjectMapper objectMapper = new ObjectMapper();
+          BlobTransferPartitionMetadata metadata = objectMapper
+              .readValue(new String(contentBytes, StandardCharsets.UTF_8), BlobTransferPartitionMetadata.class);
+          LOGGER.info("Metadata transfer completed, topic {} metadata is {}", payload.getTopicName(), metadata);
+          return metadata;
+        }
+      }
+    }
+    return null;
```

Comment:
Should we even allow `null` happening for metadata transfer?

Reply:
Yes, we need. as handleMetadataTransferContent return value need to be either null or transferred metadata. When last http content receive, return metadata, otherwise it should be null. And in updateStorePartitionMetadata, if metadata is null, it will throw exception.

Reply:
Hmm..I prefer we should just throw exception when there's any issue to retrieve the metadata rather than waiting until the very end, but I suppose that's fine